s3-fallback-store/src/main/scala/akka/persistence/s3fallback/S3Fallback.scala

```diff
@@ -0,0 +1,529 @@
+        } else {
+          Future.failed(new IllegalStateException("No Akka snapshot metadata found"))
+        }
+      }
+      .andThen {
+        case Success(_) =>
+          if (log.isDebugEnabled) {
+            log.debug(s"Loaded snapshot from bucket [$bucket] for persistenceId [$pid] seqNr [$seqNr]")
+          }
+
+        case Failure(ex) =>
+          log.error(
+            "Failed to load snapshot from bucket [{}] for persistenceId [{}] seqNr [{}]",
+            bucket,
+            pid,
+            seqNr,
+            ex)
+      }
+      .recover { case _: NoSuchKeyException | _: NoSuchBucketException =>
+        None
+      }(ExecutionContext.parasitic)
+  }
+
+  def saveSnapshot(
+      persistenceId: String,
+      seqNr: Long,
+      writeTimestamp: Instant,
+      eventTimestamp: Instant,
+      serId: Int,
+      serManifest: String,
+      payload: Array[Byte],
+      tags: Set[String],
+      meta: Option[((Int, String), Array[Byte])]): Future[String] =
+    snapshotFallbackInitialFuture.flatMap { _ =>
+      _saveSnapshot(persistenceId, seqNr, writeTimestamp, eventTimestamp, serId, serManifest, payload, tags, meta)
+    }
+
+  private def _saveSnapshot(
+      persistenceId: String,
+      seqNr: Long,
+      writeTimestamp: Instant,
+      eventTimestamp: Instant,
+      serId: Int,
+      serManifest: String,
+      payload: Array[Byte],
+      tags: Set[String],
+      meta: Option[((Int, String), Array[Byte])]): Future[String] = {
+    val bucket = settings.snapshotsBucket
+
+    if (log.isDebugEnabled) {
+      log.debug(
+        s"Saving snapshot to bucket [$bucket] for persistenceId [$persistenceId] (seqNr [$seqNr]) of size [${payload.length}] bytes")
+    }
+
+    val baseMeta = Map(
+      "akka-entity-seq-nr" -> seqNr.toString,
+      "akka-write-timestamp" -> writeTimestamp.toString,
+      "akka-event-timestamp" -> eventTimestamp.toString,
+      "akka-ser-id" -> serId.toString,
+      "akka-ser-manifest" -> serManifest.toString)
+
+    val withAkkaMeta = meta.fold(baseMeta) { case ((serId, manifest), payload) =>
+      baseMeta ++ Map(
+        "akka-meta-ser-id" -> serId.toString,
+        "akka-meta-ser-manifest" -> serManifest,
+        "akka-meta-payload" -> Base62.encode(payload))
+    }
+
+    val sortedTags = tags.toSeq.sorted.iterator.zipWithIndex.map { case (tag, idx) =>
+      s"akka-tag-$idx" -> tag
+    }
+
+    val req = PutObjectRequest.builder
+      .bucket(bucket)
+      .key(keyForPid(persistenceId))
+      .metadata((withAkkaMeta ++ sortedTags).asJava)
+      .contentType("application/octet-stream")
+      .build()
+
+    client
+      .putObject(req, AsyncRequestBody.fromBytes(payload))
+      .asScala
+      .andThen {
+        case Success(_) =>
+          if (log.isDebugEnabled) {
+            log.debug(
+              s"Saved snapshot to bucket [$bucket] for persistenceId [$persistenceId] (seqNr [$seqNr])",
+              bucket,
+              persistenceId,
+              seqNr)
```

Comment:
Using mix of string interpolation and passing args without templated format message:

```suggestion
            log.debug(
              "Saved snapshot to bucket [{}] for persistenceId [{}] (seqNr [{}])",
              bucket,
              persistenceId,
              seqNr)
```